ratelimit/redis.go

```diff
@@ -120,78 +117,70 @@ func (c *clusterLimitRedis) startSpan(ctx context.Context, spanName string) func
 //
 // If a context is provided, it uses it for creating an OpenTracing span.
 func (c *clusterLimitRedis) AllowContext(ctx context.Context, clearText string) bool {
-	s := getHashedKey(clearText)
 	c.metrics.IncCounter(redisMetricsPrefix + "total")
+	now := time.Now()
+	finishSpan := c.startSpan(ctx, allowSpanName)
+
+	allow, err := c.allowContext(ctx, clearText)
+
+	finishSpan(err != nil)
+	queryFailure := err != nil
+	c.measureQuery(allowMetricsFormat, allowMetricsFormatWithGroup, &queryFailure, now)
+
+	if err != nil {
+		allow = true // fail open
+	}
+	if allow {
+		c.metrics.IncCounter(redisMetricsPrefix + "allows")
+	} else {
+		c.metrics.IncCounter(redisMetricsPrefix + "forbids")
+	}
+	return allow
+}
+
+func (c *clusterLimitRedis) allowContext(ctx context.Context, clearText string) (bool, error) {
+	s := getHashedKey(clearText)
 	key := c.prefixKey(s)
 
 	now := time.Now()
-	var queryFailure bool
-	defer c.measureQuery(allowMetricsFormat, allowMetricsFormatWithGroup, &queryFailure, now)
-
 	nowNanos := now.UnixNano()
 	clearBefore := now.Add(-c.window).UnixNano()
 
-	count, err := c.allowCheckCard(ctx, key, clearBefore)
+	// drop all elements of the set which occurred before one interval ago.
+	_, err := c.ringClient.ZRemRangeByScore(ctx, key, 0.0, float64(clearBefore))
 	if err != nil {
-		log.Errorf("Failed to get from redis cardinality: %v", err)
-		queryFailure = true
-		// we don't return here, as we still want to record the request with ZAdd, but we mark it as a
-		// failure for the metrics
```

Comment:
If Redis op failed the subsequent will likely fail as well so it makes not sense to complicate logic here and it is better just to return an error.

Reply:
This sounds reasonable, however, I don't know how ZRemRangeByScore behaves when a key is missing, e.g. when it's a new key or the selected redis instance has just restarted. If that happens, then we would never call ZAdd

Reply:
https://redis.io/commands/zremrangebyscore returns the number of elements removed.
It seems to return 0 for non-existing zset, not an error:
```
127.0.0.1:6379> ZREMRANGEBYSCORE doesnotexist -inf 10
(integer) 0
```
Besides we should have failed tests that check denial if we never add a key.

Reply:
:+1: